util/argo/diff.go

```diff
@@ -0,0 +1,379 @@
+	diffConfig *diffConfig
+}
+
+// NewDiffConfigBuilder create a new DiffConfigBuilder instance.
+func NewDiffConfigBuilder() *DiffConfigBuilder {
+	return &DiffConfigBuilder{
+		diffConfig: &diffConfig{},
+	}
+}
+
+// WithIgnores sets the list of ResourceIgnoreDifferences in the diff config. Will set an
+// empty list if the input parameter is nil.
+func (b *DiffConfigBuilder) WithIgnores(i []v1alpha1.ResourceIgnoreDifferences) *DiffConfigBuilder {
+	ignores := i
+	if ignores == nil {
+		ignores = []v1alpha1.ResourceIgnoreDifferences{}
+	}
+	b.diffConfig.ignores = ignores
+	return b
+}
+
+// WithOverrides sets the map of ResourceOverride in the diff config. Will set an
+// empty map if the input parameter is nil.
+func (b *DiffConfigBuilder) WithOverrides(o map[string]v1alpha1.ResourceOverride) *DiffConfigBuilder {
+	overrides := o
+	if overrides == nil {
+		overrides = make(map[string]appv1.ResourceOverride)
+	}
+	b.diffConfig.overrides = overrides
+	return b
+}
+
+// WithAppLabelKey sets the appLabelKey in the diff config.
+func (b *DiffConfigBuilder) WithAppLabelKey(l string) *DiffConfigBuilder {
+	b.diffConfig.appLabelKey = &l
+	return b
+}
+
+// WithTrackingMethod sets the trackingMethod in the diff config.
+func (b *DiffConfigBuilder) WithTrackingMethod(t string) *DiffConfigBuilder {
+	b.diffConfig.trackingMethod = &t
+	return b
+}
+
+// WithAppName sets the appName in the diff config. The appName is only required
+// if retrieving the diff from the cache (NoCache = false).
+func (b *DiffConfigBuilder) WithAppName(n string) *DiffConfigBuilder {
+	b.diffConfig.appName = &n
+	return b
+}
+
+// WithNoCache sets the nocache in the diff config. Defines if it should retrieve
+// the computed diff from the cache.
+func (b *DiffConfigBuilder) WithNoCache(c bool) *DiffConfigBuilder {
+	b.diffConfig.noCache = &c
+	return b
+}
+
+// WithStateCache sets the appstatecache.Cache in the diff config. Only required if
+// retrieving the diff from the cache (NoCache = false).
+func (b *DiffConfigBuilder) WithStateCache(s *appstatecache.Cache) *DiffConfigBuilder {
+	b.diffConfig.stateCache = s
+	return b
+}
+
+// WithIgnoreAggregatedRoles sets the ignoreAggregatedRoles in the diff config.
+func (b *DiffConfigBuilder) WithIgnoreAggregatedRoles(i bool) *DiffConfigBuilder {
+	b.diffConfig.ignoreAggregatedRoles = &i
+	return b
+}
+
+// WithLogger sets the logger in the diff config.
+func (b *DiffConfigBuilder) WithLogger(l logr.Logger) *DiffConfigBuilder {
+	b.diffConfig.logger = l
+	return b
+}
+
+// Build will first validate the current state of the diff config and return the
+// DiffConfig implementation if no errors are found. Will return nil and the error
+// details otherwise.
+func (b *DiffConfigBuilder) Build() (DiffConfig, error) {
+	err := b.diffConfig.Validate()
+	if err != nil {
+		return nil, err
+	}
+	return b.diffConfig, nil
+}
+
+// DiffConfig defines methods to retrieve the configurations used while applying diffs.
+type DiffConfig interface {
```

Comment:
Nitpick: we could just avoid introducing an interface and instead introduce `DataConfig` struct which has only private fields. This still ensures that consumer cannot modify config fields

Reply:
I thought about that but I think that with interface is safer. The reason is because consumers would still be able to instantiate an empty `DataConfig` struct and pass that as a parameter without validations. With the interface, the only way to get a valid instance is calling `DiffConfigBuilder.Build` which runs all the validations before returning the instance.